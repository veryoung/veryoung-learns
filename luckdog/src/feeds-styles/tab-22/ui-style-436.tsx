/**
 * 新浅用户福利卡
 */
import React, { ForwardRefRenderFunction, useEffect, useState } from 'react';
import { View, StyleSheet, Image, Text, ImageBackground } from '@tencent/hippy-react-qb';
import { FormatUtils } from '../common/utils';
import { strictExposeReporter, reportUDS, BusiKey, logError } from '@/luckdog';
import { WelfareCountDown } from '../../components/welfare-count-down';
import AwardRecordCarousel from '../../components/lottery-record-carousel';
import { showToast } from '@/luckbox';
import FeedsIcon from '../../framework/FeedsIcon';
import FeedsTheme from '../../framework/FeedsTheme';
import { FeedsUIStyle, CardRadius } from '../../framework/FeedsConst';
import { ViewRef } from '@/entity';


const TAG = 'ui-style-436';

/** 内容区域高度 */
const CONTENT_HEIGHT = 150;

const styles = StyleSheet.create({
  wrap: {
    overflow: 'hidden',
  },
  // 背景图片加载失败时的兼容样式
  imageErrorWrap: {
    height: CONTENT_HEIGHT,
    marginTop: FeedsUIStyle.FEEDS_CARD_MARGIN_VERTICAL,
  },
  image: {
    flex: 1,
    resizeMode: 'cover',
  },
  content: {
    marginLeft: 12,
    marginRight: 12,
    backgroundColor: '#997649',
    borderRadius: CardRadius,
  },
  contentBg: {
    height: CONTENT_HEIGHT,
    justifyContent: 'space-between',
    resizeMode: 'cover',
    paddingTop: 16,
    paddingLeft: 14,
    paddingRight: 14,
    borderRadius: CardRadius,
    overflow: 'hidden',
  },
  top: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    height: 20,
  },
  awardList: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  bigImage: {
    resizeMode: 'contain',
    width: 93,
    height: 100,
    marginLeft: 2,
  },
  award: {
    alignItems: 'center',
  },
  awardImage: {
    width: 50,
    height: 50,
  },
  awardNameWrap: {
    marginTop: 4,
    paddingTop: 2.5,
    paddingLeft: 8,
    paddingRight: 8,
    paddingBottom: 2.5,
    height: 20,
    borderRadius: 10,
    justifyContent: 'center',
  },
  awardName: {
    fontSize: 11,
    colors: FeedsTheme.SkinColor.D2_1,
    textAlign: 'center',
  },
});

const onLayout = (event: Record<string, any>, cardIndex: number): void => {
  strictExposeReporter.addExpoItem({
    cardIndex,
    tabIndex: 0,
    bookIndex: 0,
    bookId: '0',
    rect: event.layout,
  });
};

const onClick = (props: Record<string, any>): void => {
  showToast('读任意小说领福利');
  reportUDS(BusiKey.CLICK__CARD, props);

  // 页面滚动到下一张卡片
  try {
    props.parent.refs.feedsList.scrollToIndex(0, props.index + 1, true);
  } catch (err) {
    logError(err, `${TAG}.onClick`);
  }
};

type Props = {
  itemBean: {
    parsedObject: {
      sData: string;
    }
  },
  index: number;
};

const NewUserWelfareCard: ForwardRefRenderFunction<ViewRef, Props> = (
  props,
  ref: React.Ref<ViewRef>,
): JSX.Element | null => {
  const {
    remainTime = 0, bgImage, bgImageShowRatio, awardRecordList, awardList,
  } = JSON.parse(props.itemBean?.parsedObject?.sData || '{}');

  if (remainTime <= 0 || !awardList?.length) {
    logError(`福利卡数据异常：${props.itemBean?.parsedObject?.sData}`, TAG);
    return null;
  }

  const [imageHeight, setImageHeight] = useState(0);
  const [height, setHeight] = useState(0);

  useEffect(() => {
    const fetchImageSize = async () => {
      try {
        const size = await Image.getSize(bgImage);
        const height = FormatUtils.formatDesignLength(size.height);
        setImageHeight(height);

        // 卡片的高度 = 图片露出的高度 + 内容的高度
        const cardHeight = (height * bgImageShowRatio) + CONTENT_HEIGHT;
        setHeight(cardHeight);
      } catch (err) {
        logError(err, `${TAG}.fetchImageSize`);
      }
    };

    fetchImageSize();
  }, [bgImage]);

  return (<View
    ref={ref}
    onLayout={event => onLayout(event, props.index)}
    style={[styles.wrap, height === 0 ? styles.imageErrorWrap : { height }]}
    onClick={() => onClick(props)}
  >
    <Image source={{ uri: bgImage }} style={[styles.image, {
      height: imageHeight,
    }]} />
    <View style={[styles.content, { marginTop: - imageHeight * bgImageShowRatio }]}>
      <ImageBackground source={{ uri: FeedsIcon.welfareCardBg }} style={styles.contentBg} >
        <View style={styles.top}>
          { awardRecordList?.length > 0 ? <AwardRecordCarousel awardRecordList={awardRecordList} /> : <View /> }
          <WelfareCountDown remainTime={remainTime} onTimeEnd={() => setHeight(0)} />
        </View>
        <View style={styles.awardList}>
          { awardList.map((award, index) => (index === 0
            ? <Image style={styles.bigImage} source={{ uri: award.imageUrl }} key={index} />
            : (<View style={styles.award} key={index}>
              <Image style={styles.awardImage} source={{ uri: award.imageUrl }} />
              <ImageBackground source={{ uri: FeedsIcon.awardBtnBg }} style={styles.awardNameWrap}>
                <Text style={styles.awardName}>{award.name}</Text>
              </ImageBackground>
            </View>)
          ))}
        </View>
      </ImageBackground>
    </View>
  </View>);
};

export default React.forwardRef(NewUserWelfareCard);

