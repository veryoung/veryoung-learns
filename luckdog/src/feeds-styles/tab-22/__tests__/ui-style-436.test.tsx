import React from 'react';
import renderer from 'react-test-renderer';

import FeedsViewUIStyle436 from '../ui-style-436';
import { itemBean } from './mocks/ui-style-436';

describe('<FeedsViewUIStyle436 />', () => {
  describe('render testing', () => {
    it('render welfare card', () => {
      const props = {
        itemBean,
        index: 0,
      };
      const tree = renderer.create(<FeedsViewUIStyle436 {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
