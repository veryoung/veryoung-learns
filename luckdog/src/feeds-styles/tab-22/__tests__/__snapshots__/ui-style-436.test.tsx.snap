// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FeedsViewUIStyle436 /> render testing render welfare card 1`] = `
<div
  nativeName="View"
  onClick={[Function]}
  onLayout={[Function]}
  style={
    Array [
      Object {
        "overflow": "hidden",
      },
      Object {
        "height": 150,
        "marginTop": 16,
      },
    ]
  }
>
  <img
    alt=""
    nativeName="Image"
    src="https://novel.imqq.com/ricoxiao/1621412327/bg_back_welfare.png"
    style={
      Object {
        "0": Object {
          "flex": 1,
          "resizeMode": "cover",
        },
        "1": Object {
          "height": 0,
        },
      }
    }
  />
  <div
    nativeName="View"
    style={
      Array [
        Object {
          "backgroundColor": "#997649",
          "borderRadius": 12,
          "marginLeft": 12,
          "marginRight": 12,
        },
        Object {
          "marginTop": -0,
        },
      ]
    }
  >
    <div
      nativeName="View"
      style={
        Object {
          "borderRadius": 12,
          "height": 150,
          "justifyContent": "space-between",
          "overflow": "hidden",
          "paddingLeft": 14,
          "paddingRight": 14,
          "paddingTop": 16,
          "resizeMode": "cover",
        }
      }
    >
      <img
        alt=""
        nativeName="Image"
        src="https://novel.imqq.com/ricoxiao/1621436847/welfare-card-bg.png"
        style={
          Array [
            Object {
              "bottom": 0,
              "height": 150,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "width": undefined,
            },
            undefined,
          ]
        }
      />
      <div
        nativeName="View"
        style={
          Object {
            "flexDirection": "row",
            "height": 20,
            "justifyContent": "space-between",
          }
        }
      >
        <div
          nativeName="View"
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFFFFF20",
              "borderRadius": 10,
              "flexDirection": "row",
              "height": 20,
              "width": 154,
            }
          }
        >
          <img
            alt=""
            nativeName="Image"
            src="https://thirdqq.qlogo.cn/g?b=sdk&k=KFoALoWw8z4ABic1ypusOQQ&s=640&t=1585262294"
            style={
              Object {
                "borderRadius": 8,
                "height": 16,
                "marginLeft": 2,
                "width": 16,
              }
            }
          />
          <p
            nativeName="Text"
            numberOfLines={1}
            style={
              Object {
                "colors": Array [
                  "rgba(0, 0, 0, 0.5)",
                  "#635647",
                  "rgba(0, 0, 0, 0.5)",
                  "#635647",
                ],
                "fontSize": 12,
                "lineHeight": 20,
                "marginLeft": 4,
                "maxWidth": 130,
                "paddingRight": 2,
              }
            }
            text="♥*♠ 抽中王者礼包"
          >
            ♥*♠ 抽中王者礼包
          </p>
        </div>
        <div
          nativeName="View"
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <p
            nativeName="Text"
            style={
              Object {
                "colors": Array [
                  "rgba(0, 0, 0, 0.5)",
                  "#635647",
                  "rgba(0, 0, 0, 0.5)",
                  "#635647",
                ],
                "fontSize": 12,
                "lineHeight": 20,
                "marginRight": 4,
              }
            }
            text="限时"
          >
            限时
          </p>
          <div
            nativeName="View"
            style={
              Array [
                Object {
                  "flexDirection": "row",
                  "justifyContent": "center",
                  "width": 80,
                },
                Object {},
              ]
            }
          >
            <p
              nativeName="Text"
              style={
                Object {
                  "backgroundColors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "borderRadius": 4,
                  "colors": Array [
                    "#000000",
                    "#686D74",
                    "#000000",
                    "#ffffff",
                  ],
                  "fontFamily": "DINNextLTPro-Medium",
                  "fontSize": 12,
                  "fontWeight": "bold",
                  "height": 20,
                  "lineHeight": 19,
                  "overflow": "hidden",
                  "textAlign": "center",
                  "width": 18,
                }
              }
              text="14"
            >
              14
            </p>
            <div
              nativeName="View"
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "width": 8,
                }
              }
            >
              <div
                nativeName="View"
                style={
                  Array [
                    Object {
                      "backgroundColors": Array [
                        "#000000",
                        "#686D74",
                        "#000000",
                        "#ffffff",
                      ],
                      "borderRadius": 1,
                      "height": 2,
                      "marginTop": 2,
                      "width": 2,
                    },
                    Object {
                      "marginTop": 0,
                    },
                  ]
                }
              />
              <div
                nativeName="View"
                style={
                  Array [
                    Object {
                      "backgroundColors": Array [
                        "#000000",
                        "#686D74",
                        "#000000",
                        "#ffffff",
                      ],
                      "borderRadius": 1,
                      "height": 2,
                      "marginTop": 2,
                      "width": 2,
                    },
                    Object {
                      "marginTop": 2,
                    },
                  ]
                }
              />
            </div>
            <p
              nativeName="Text"
              style={
                Object {
                  "backgroundColors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "borderRadius": 4,
                  "colors": Array [
                    "#000000",
                    "#686D74",
                    "#000000",
                    "#ffffff",
                  ],
                  "fontFamily": "DINNextLTPro-Medium",
                  "fontSize": 12,
                  "fontWeight": "bold",
                  "height": 20,
                  "lineHeight": 19,
                  "overflow": "hidden",
                  "textAlign": "center",
                  "width": 18,
                }
              }
              text="36"
            >
              36
            </p>
            <div
              nativeName="View"
              style={
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                  "width": 8,
                }
              }
            >
              <div
                nativeName="View"
                style={
                  Array [
                    Object {
                      "backgroundColors": Array [
                        "#000000",
                        "#686D74",
                        "#000000",
                        "#ffffff",
                      ],
                      "borderRadius": 1,
                      "height": 2,
                      "marginTop": 2,
                      "width": 2,
                    },
                    Object {
                      "marginTop": 0,
                    },
                  ]
                }
              />
              <div
                nativeName="View"
                style={
                  Array [
                    Object {
                      "backgroundColors": Array [
                        "#000000",
                        "#686D74",
                        "#000000",
                        "#ffffff",
                      ],
                      "borderRadius": 1,
                      "height": 2,
                      "marginTop": 2,
                      "width": 2,
                    },
                    Object {
                      "marginTop": 2,
                    },
                  ]
                }
              />
            </div>
            <p
              nativeName="Text"
              style={
                Object {
                  "backgroundColors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "borderRadius": 4,
                  "colors": Array [
                    "#000000",
                    "#686D74",
                    "#000000",
                    "#ffffff",
                  ],
                  "fontFamily": "DINNextLTPro-Medium",
                  "fontSize": 12,
                  "fontWeight": "bold",
                  "height": 20,
                  "lineHeight": 19,
                  "overflow": "hidden",
                  "textAlign": "center",
                  "width": 18,
                }
              }
              text="30"
            >
              30
            </p>
          </div>
        </div>
      </div>
      <div
        nativeName="View"
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "space-between",
          }
        }
      >
        <img
          alt=""
          nativeName="Image"
          src="https://novel.imqq.com/ricoxiao/1621863954/tencent_vip_big.png"
          style={
            Object {
              "height": 100,
              "marginLeft": 2,
              "resizeMode": "contain",
              "width": 93,
            }
          }
        />
        <div
          nativeName="View"
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <img
            alt=""
            nativeName="Image"
            src="https://novel.imqq.com/ricoxiao/1621863954/coin.png"
            style={
              Object {
                "height": 50,
                "width": 50,
              }
            }
          />
          <div
            nativeName="View"
            style={
              Object {
                "borderRadius": 10,
                "height": 20,
                "justifyContent": "center",
                "marginTop": 4,
                "paddingBottom": 2.5,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 2.5,
              }
            }
          >
            <img
              alt=""
              nativeName="Image"
              src="https://novel.imqq.com/ricoxiao/1621949413/award_btn_bg.png"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "height": 20,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
            <p
              nativeName="Text"
              style={
                Object {
                  "colors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "fontSize": 11,
                  "textAlign": "center",
                }
              }
              text="快乐金币"
            >
              快乐金币
            </p>
          </div>
        </div>
        <div
          nativeName="View"
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <img
            alt=""
            nativeName="Image"
            src="https://novel.imqq.com/ricoxiao/1621863954/lottery_package.png"
            style={
              Object {
                "height": 50,
                "width": 50,
              }
            }
          />
          <div
            nativeName="View"
            style={
              Object {
                "borderRadius": 10,
                "height": 20,
                "justifyContent": "center",
                "marginTop": 4,
                "paddingBottom": 2.5,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 2.5,
              }
            }
          >
            <img
              alt=""
              nativeName="Image"
              src="https://novel.imqq.com/ricoxiao/1621949413/award_btn_bg.png"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "height": 20,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
            <p
              nativeName="Text"
              style={
                Object {
                  "colors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "fontSize": 11,
                  "textAlign": "center",
                }
              }
              text="王者礼包"
            >
              王者礼包
            </p>
          </div>
        </div>
        <div
          nativeName="View"
          style={
            Object {
              "alignItems": "center",
            }
          }
        >
          <img
            alt=""
            nativeName="Image"
            src="https://novel.imqq.com/ricoxiao/1621863954/magic_gift.png"
            style={
              Object {
                "height": 50,
                "width": 50,
              }
            }
          />
          <div
            nativeName="View"
            style={
              Object {
                "borderRadius": 10,
                "height": 20,
                "justifyContent": "center",
                "marginTop": 4,
                "paddingBottom": 2.5,
                "paddingLeft": 8,
                "paddingRight": 8,
                "paddingTop": 2.5,
              }
            }
          >
            <img
              alt=""
              nativeName="Image"
              src="https://novel.imqq.com/ricoxiao/1621949413/award_btn_bg.png"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "height": 20,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 0,
                    "width": undefined,
                  },
                  undefined,
                ]
              }
            />
            <p
              nativeName="Text"
              style={
                Object {
                  "colors": Array [
                    "#ffffffff",
                    "#31363bff",
                    "#ffffffff",
                    "#31363bff",
                  ],
                  "fontSize": 11,
                  "textAlign": "center",
                }
              }
              text="神秘礼包"
            >
              神秘礼包
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
